test/e2e/clusterclass_rollout.go

```diff
@@ -0,0 +1,1094 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog/v2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/cluster-api/controllers/external"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/internal/contract"
+	"sigs.k8s.io/cluster-api/internal/controllers/topology/machineset"
+	"sigs.k8s.io/cluster-api/test/e2e/internal/log"
+	"sigs.k8s.io/cluster-api/test/framework"
+	"sigs.k8s.io/cluster-api/test/framework/clusterctl"
+	"sigs.k8s.io/cluster-api/util"
+	"sigs.k8s.io/cluster-api/util/patch"
+)
+
+// ClusterClassRolloutSpecInput is the input for ClusterClassRolloutSpec.
+type ClusterClassRolloutSpecInput struct {
+	E2EConfig             *clusterctl.E2EConfig
+	ClusterctlConfigPath  string
+	BootstrapClusterProxy framework.ClusterProxy
+	ArtifactFolder        string
+	SkipCleanup           bool
+	ControlPlaneWaiters   clusterctl.ControlPlaneWaiters
+
+	// Flavor is the cluster-template flavor used to create the Cluster for testing.
+	// NOTE: The template must be using ClusterClass, KCP and CABPK as this test is specifically
+	// testing ClusterClass and KCP rollout behavior.
+	Flavor string
+}
+
+// ClusterClassRolloutSpec implements a test that verifies the ClusterClass rollout behavior.
+// It specifically covers the in-place propagation behavior from ClusterClass / Cluster topology to all
+// objects of the Cluster topology (e.g. KCP, MD) and even tests label propagation to the Nodes of the
+// workload cluster.
+// Thus, the test consists of the following steps:
+//   - Deploy Cluster using a ClusterClass and wait until it is fully provisioned.
+//   - Assert cluster objects
+//   - Modify in-place mutable fields of KCP and the MachineDeployments
+//   - Verify that fields were mutated in-place and assert cluster objects
+//   - Modify fields in KCP and MachineDeployments to trigger a full rollout of all Machines
+//   - Verify that all Machines have been replaced and assert cluster objects
+//   - Set RolloutAfter on KCP and MachineDeployments to trigger a full rollout of all Machines
+//   - Verify that all Machines have been replaced and assert cluster objects
```

Comment:
Not for now, but I think that if we expand this a little bit this part we can merge ClusterClassChangesSpec into this test

Reply:
I was thinking about the overlap between those tests as well.

Currently I went with separate tests as the ClusterClass changes test can be used with arbitrary control plane and bootstrap providers while the rollout test specifically tests behavior of KCP (in-place propagation).

As in-place propagation is not part of our contract and not even all control plane providers have a machineTemplate I thought it to be cleaner / easier to understand if we have a clean separation. So that one test is testing basic ClusterClass "actions" while the other one is testing the entire in-place propagation specifically for KCP etc.

I think there is probably a way to add some flags etc so we can have both in one test. I'm just not sure if this makes the test to complicated. Right now both tests seem nicely focused on their task, although there is definitely overlap.

WDYT?

Reply:
No rush, let's just keep it into the radar